File: bindings/julia/test/test_database_describe.jl
module TestDatabaseDescribe

using Quiver
using Test

include("fixture.jl")

function open_db()
    path_schema = joinpath(tests_path(), "schemas", "valid", "collections.sql")
    return Quiver.from_schema(":memory:", path_schema)
end

# Shared `tests/schemas/ui/` fixture corpus (CORPUS-03) -- never copied into this binding.
ui_fixture_path(name) = joinpath(tests_path(), "schemas", "ui", name)

function open_ui_fixture(name, stem)
    dir = ui_fixture_path(name)
    return Quiver.from_schema(joinpath(dir, "julia_$(stem).sqlite"), joinpath(dir, "schema.sql"))
end

@testset "Describe" begin
    @testset "describe" begin
        db = open_db()
        @test Quiver.describe(db) isa String
        Quiver.close!(db)
    end

    @testset "describe_collection" begin
        db = open_db()
        @test Quiver.describe_collection(db, "Collection") isa String
        Quiver.close!(db)
    end

    @testset "summarize_collection" begin
        db = open_db()
        @test Quiver.summarize_collection(db, "Collection") isa String
        Quiver.close!(db)
    end

    # DESC-07: exact-string enum rendering against the shared tests/schemas/ui/ fixtures --
    # a "returns a String" assertion alone cannot catch a per-binding decoding bug (D-31).

    @testset "enum vocabulary declared with zero elements" begin
        db = open_ui_fixture("enum_basic", "declared")
        report = Quiver.describe_collection(db, "Storage")
        @test occursin("enum bool {0: Disabled, 1: Enabled}", report)
        Quiver.close!(db)
    end

    @testset "enum histogram over twelve elements" begin
        db = open_ui_fixture("enum_basic", "histogram")
        for i in 1:8
            Quiver.create_element!(db, "Storage"; label = "Disabled $(i)", has_commitment = 0)
        end
        for i in 1:4
            Quiver.create_element!(db, "Storage"; label = "Enabled $(i)", has_commitment = 1)
        end
        report = Quiver.summarize_collection(db, "Storage")
        @test occursin("values {0: 8 (Disabled), 1: 4 (Enabled)}", report)
        Quiver.close!(db)
    end

    @testset "UI config header line" begin
        db = open_ui_fixture("enum_basic", "header")
        report = Quiver.describe(db)
        @test occursin("UI config: ", report)
        @test occursin(" (locale: en)", report)
        Quiver.close!(db)
    end

    @testset "unit and hidden decoration" begin
        db = open_ui_fixture("enum_basic", "unit_hidden")
        report = Quiver.describe_collection(db, "Storage")
        @test occursin("[MW]", report)
        @test occursin("[hidden]", report)
        Quiver.close!(db)
    end

    @testset "accented enum label renders byte-for-byte" begin
        db = open_ui_fixture("foresight_like", "accented")
        report = Quiver.describe_collection(db, "EconomicDriver")
        @test occursin("Seasonal Naïve", report)
        Quiver.close!(db)
    end

    @testset "no ui/ directory leaves the header off" begin
        db = open_ui_fixture("no_ui_dir", "no_sidecar")
        report = Quiver.describe(db)
        @test !occursin("UI config: ", report)
        Quiver.close!(db)
    end
end

end

File: bindings/python/tests/test_database_describe.py
from __future__ import annotations

from collections.abc import Generator
from pathlib import Path

import pytest

from quiverdb import Database


# DESC-07: exact-string enum rendering against the shared tests/schemas/ui/ fixtures -- a
# "returns a str" assertion alone (kept in test_database_metadata.py's TestDescribe classes)
# cannot catch a per-binding decoding bug (D-31).
#
# Deliberately NOT built on the existing pytest temp-directory fixtures: a database created in
# a bare temp directory has no `ui/` sibling directory, so the test would pass by silently
# reading no sidecar at all and asserting the old, unchanged output. Each database here is
# built inside the fixture directory itself, under a filename unique to this binding and this
# test, so it cannot collide with the other bindings' or the C++/C API suites' own fixture
# databases.


@pytest.fixture
def ui_fixture(schemas_path: Path):
    def _ui_fixture(name: str) -> Path:
        return schemas_path / "ui" / name

    return _ui_fixture


def _open_ui_fixture(ui_fixture, name: str, stem: str) -> Database:
    directory = ui_fixture(name)
    return Database.from_schema(str(directory / f"python_{stem}.sqlite"), str(directory / "schema.sql"))


@pytest.fixture
def enum_declared_db(ui_fixture) -> Generator[Database, None, None]:
    database = _open_ui_fixture(ui_fixture, "enum_basic", "declared")
    yield database
    database.close()


@pytest.fixture
def enum_histogram_db(ui_fixture) -> Generator[Database, None, None]:
    database = _open_ui_fixture(ui_fixture, "enum_basic", "histogram")
    yield database
    database.close()


@pytest.fixture
def enum_header_db(ui_fixture) -> Generator[Database, None, None]:
    database = _open_ui_fixture(ui_fixture, "enum_basic", "header")
    yield database
    database.close()


@pytest.fixture
def enum_unit_hidden_db(ui_fixture) -> Generator[Database, None, None]:
    database = _open_ui_fixture(ui_fixture, "enum_basic", "unit_hidden")
    yield database
    database.close()


@pytest.fixture
def foresight_accented_db(ui_fixture) -> Generator[Database, None, None]:
    database = _open_ui_fixture(ui_fixture, "foresight_like", "accented")
    yield database
    database.close()


@pytest.fixture
def no_sidecar_db(ui_fixture) -> Generator[Database, None, None]:
    database = _open_ui_fixture(ui_fixture, "no_ui_dir", "no_sidecar")
    yield database
    database.close()


class TestEnumRendering:
    def test_declared_vocabulary_with_zero_elements(self, enum_declared_db: Database) -> None:
        report = enum_declared_db.describe_collection("Storage")
        assert "enum bool {0: Disabled, 1: Enabled}" in report

    def test_histogram_over_twelve_elements(self, enum_histogram_db: Database) -> None:
        for i in range(8):
            enum_histogram_db.create_element("Storage", label=f"Disabled {i}", has_commitment=0)
        for i in range(4):
            enum_histogram_db.create_element("Storage", label=f"Enabled {i}", has_commitment=1)

        report = enum_histogram_db.summarize_collection("Storage")
        assert "values {0: 8 (Disabled), 1: 4 (Enabled)}" in report

    def test_ui_config_header_line(self, enum_header_db: Database) -> None:
        report = enum_header_db.describe()
        assert "UI config: " in report
        assert " (locale: en)" in report

    def test_unit_and_hidden_decoration(self, enum_unit_hidden_db: Database) -> None:
        report = enum_unit_hidden_db.describe_collection("Storage")
        assert "[MW]" in report
        assert "[hidden]" in report

    def test_accented_enum_label_renders_byte_for_byte(self, foresight_accented_db: Database) -> None:
        report = foresight_accented_db.describe_collection("EconomicDriver")
        assert "Seasonal Naïve" in report

    def test_no_ui_directory_leaves_header_off(self, no_sidecar_db: Database) -> None:
        report = no_sidecar_db.describe()
        assert "UI config: " not in report
